### PR TITLE
Support stacking of environments

### PR DESCRIPTION
**Pull request by abretaud**
Hi,
This PR is a follow up to #3580 where there was some discussion about allowing (or not) people to *stack* multiple environments.
I thought it would be easier to discuss it with real code, so here's a first (no tested!) proposition: it adds a --stack option to the activate script, that the user must add for each environment he wishes to stack.
Before writing any other code (for other shells, doc, etc); I would like to get some feedback on it: do you think something like this could be merged?
Any suggestion is of course welcome

### CODE REVIEW

**Comment by kalefranz** [comment]
Thanks for the PR @abretaud.  At a minimum it should keep the conversation going.

File: shell/activate.sh [hunk 1]
```diff
@@ -174,23 +190,33 @@ fi
...
+        if [ -n "${_MSYS2_ENV_CONV_EXCL+x}" ]; then
+            unset _MSYS2_ENV_CONV_EXCL
+        fi
+        return 1
+    fi
+else
+    if [ "${CONDA_VERBOSE}" = "${TRUE}" ]; then
+        echo "[ACTIVATE]: Warning: stacking environments (--stacks option) can cause unexpected errors in case of incompatibilities between different conda packages"
     fi
-    return 1
```
Comment (njalerikson) [inline shell/activate.sh]: This effectively means that any --stack will detect that it's being stacked and then not deactivate the current environment and then exits without activating the new environment to stack.

**Comment by njalerikson** [comment]
I think that in general the bulk of this kind of logic should occur in a conda function. Further there is a lot more checks and processing that should be done beyond not deactivating existing environments.

And we would need to rethink how all of the backup variables are dealt with. Like how do we update the prompt for stacked environments? How do we restore the prompts in a deactivate? How do we keep track of the `CONDA_PREFIX` value that is used to deactivate an environment? Etc.

Also some kind of similar logic should be added to deactivate to allow a user to unstack a specific environment.

In general I think a lot of important checks that should be implemented would best be implemented after a few other updates to conda (see the idea of a `.conda` directory discussed in #3808).

I think this is an important feature to consider as it has significant HPC value (this along with multi-user support #1329).

File: shell/activate.sh (hunk 1 above)
Comment (abretaud) [inline shell/activate.sh]: Hum, yes, I think the if/else logic was wrong, is it ok now?

**Comment by abretaud** [comment]
> I think that in general the bulk of this kind of logic should occur in a conda function. Further there is a lot more checks and processing that should be done beyond not deactivating existing environments.

I'm not sure to understand what you mean by "a conda function", is there some other code modifying the PATH apart from these activate/seactivate scripts?

> how do we update the prompt for stacked environments?

Here I just write a fixed "(stacked env)" prefix, I fear that it could be unreadable if the list of envs was listed. But maybe having access (in an env var?) to the list of loaded envs would be handy?

> How do we restore the prompts in a deactivate?
> How do we keep track of the CONDA_PREFIX value that is used to deactivate an environment? Etc.
> Also some kind of similar logic should be added to deactivate to allow a user to unstack a specific environment.

Yep, my idea was that a call to deactivate would deactivate all stacked envs at once. Deactivating each stacked env individually would be cool indeed (but not top priority for me). And I thought deactivate was just resetting the PATH to a backed up value, which is not the case apparently. I guess this should work if CONDA_PREFIX was an array?

**Comment by njalerikson** [comment]
I fully believe that implementing this idea is easy. Implementing a safe and robust solution is not "easy" as it would need to provide some basic protections to the user from making stupid stacking errors. Putting that extra work into a conda function (e.g. `conda ..checkstacking`, very much like the `conda ..changeps1`) will allow us to concisely program the solution in Python and still be very OS-agnostic and shell-agnostic. Again I am not sure how much work should go into this...

----

Correct, `deactivate` does not simply restore the `PATH` to the original. See lines ~164-180 in `deactivate.sh`, also reproduced below:

```
###########################################################################
# RESTORE PATH                                                            #
# remove only first instance of $CONDA_PREFIX from $PATH, since this is   #
# Unix we will only expect a single path that need to be removed, for     #
# simplicity and consistency with the Windows *.bat scripts we will use   #
# fuzzy matching (/f) to get all of the relevant removals                 #
PATH=$(envvar_cleanup.bash "${PATH}" --delim=":" -u -f "${CONDA_PREFIX}")
if [ $? != 0 ]; then
    [ "${IS_ENV_CONDA_VERBOSE}" = "${FALSE}" ] && unset CONDA_VERBOSE
    unset IS_ENV_CONDA_VERBOSE
    unset TRUE
    unset FALSE
    return 1
fi
export PATH
# END RESTORE PATH                                                        #
###########################################################################
```

Converting `$CONDA_PREFIX` into an array of sorts is an option but then again arrays are far from shell agnostic.

**Comment by asmeurer** [comment]
IMHO this is a broken abstraction. A huge gotcha here is that only the PATH is stacked. Packages in stacked environments don't interact with each other except inasmuch as they use the PATH. So for example, you won't be able to import Python packages from one environment using the python from another. This will lead to leaky attempts to fix the abstraction by setting more and more environment variables, such as PYTHONPATH and LD_LIBRARY_PATH, and there will always be something that does its own prefix-based searching, or a new environment variable to set. 

Conda environments have always been the superior abstraction, since things in the same environment will always interact with each other as you'd expect without any special environment variables being set. That's why I think it would be better to add some UI to make "merging" conda environments easier (meaning a new environment with merged packages from two or more other environments).

A side note: there's a separate issue to make deactivate just restore the pre-activated PATH.

**Comment by mforbes** [comment]
@asmeurer I have one use case that I don't see how a single environment can help: I would like to be able to manage mercurial and a bunch of related packages (which requires python 2) while using a python 3 environment for most of my work.  This seems to require the ability to activate two environments simultaneously or install mercurial outside of the environment.  How would you propose managing this?

**Comment by kalefranz** [comment]
This PR is superseded by #5159, which is a simple adjustment after #5044.  The activate and deactivate scripts that this PR modifies have been deprecated.

**Pull request closed without merging**